libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/IncrementalCompilationMultiProjectIT.kt

```diff
@@ -151,67 +153,109 @@ class IncrementalCompilationClasspathSnapshotJvmMultiProjectIT : IncrementalComp
 
 abstract class BaseIncrementalCompilationMultiProjectIT : IncrementalCompilationBaseIT() {
 
+    protected abstract val compileKotlinTaskName: String
+
+    protected abstract val additionalLibDependencies: String
+
+    protected val changeMethodSignatureInLib: Project.() -> Unit = {
+        File(projectDir, "lib").getFileByName("A.kt").modify {
+            it.replace("fun a() {}", "fun a(): Int = 1")
+        }
+    }
+
+    protected val changeMethodBodyInLib: Project.() -> Unit = {
+        File(projectDir, "lib").getFileByName("A.kt").modify {
+            it.replace("fun a() {}", "fun a() { println() }")
+        }
+    }
+
     @Test
     fun testAbiChangeInLib_changeMethodSignature() {
         doTest(
-            "A.kt",
-            { it.replace("fun a() {}", "fun a(): Int = 1") },
-            expectedAffectedFileNames = listOf("A.kt", "B.kt", "AA.kt", "AAA.kt", "BB.kt", "barUseA.kt", "fooUseA.kt")
+            modifyProject = changeMethodSignatureInLib,
+            expectedCompiledFileNames = listOf(
+                "A.kt", "B.kt", "barUseA.kt", // In lib
+                "AA.kt", "AAA.kt", "BB.kt", "fooUseA.kt" // In app
+            )
         )
     }
 
     @Test
     fun testAbiChangeInLib_addNewMethod() {
         doTest(
-            "A.kt",
-            { it.replace("fun a() {}", "fun a() {}\nfun newA() {}") },
-            expectedAffectedFileNames = listOf("A.kt", "B.kt", "AA.kt", "AAA.kt", "BB.kt")
+            modifyProject = {
+                File(projectDir, "lib").getFileByName("A.kt").modify {
+                    it.replace("fun a() {}", "fun a() {}\nfun newA() {}")
+                }
+            },
+            expectedCompiledFileNames = listOf(
+                "A.kt", "B.kt", // In lib
+                "AA.kt", "AAA.kt", "BB.kt" // In app
+            )
         )
     }
 
     @Test
     fun testNonAbiChangeInLib_changeMethodBody() {
         doTest(
-            "A.kt",
-            { it.replace("fun a() {}", "fun a() { println() }") },
-            expectedAffectedFileNames = listOf("A.kt")
+            modifyProject = changeMethodBodyInLib,
+            expectedCompiledFileNames = listOf("A.kt") // In lib
         )
     }
 
     @Test
-    fun testMoveFunctionFromLibToApp() {
+    fun testAddDependencyInLib() {
```

Comment:
There used to be one more incremental build in this test case, asserting that we can run have build -> (change dep) non-incremental build -> (change source) incremental build i.e that we can "recover" after a non-incremental one.  Should we restore old behavior?

Reply:
That's a good catch! I thought about that too and felt the "recovery" check is not necessary.

The reason is that a non-incremental build should be deterministic, regardless of what happened before (with the exception of cache hit with `@LocalState`, but we're not testing caching here). So if a build is non-incremental, the next build should always be incremental (given a typical change).

The "recovery" check has a tiny bit of value, but it is added to several test cases making the tests not focused and increasing test execution time.

If the check is important, I can add a separate test just for it, but it's not clear to me which scenario we should use to set up the test.

Also: When classpath snapshot is enabled, app compilation will be incremental right in the second build, so I guess this won't be an issue in the long run when the feature is enabled by default.

Reply:
SGTM, I agree there are other incremental scenario tests that cover this.